src/orange/OrangeParams.cc

```diff
@@ -147,17 +147,13 @@ OrangeParams::OrangeParams(OrangeInput input)
     ui_surf.push_back(0);
     ui_vol.push_back(0);
 
-    auto GetNumSurfaces
-        = Overload<std::function<size_type(UnitInput const&)>,
-                   std::function<size_type(RectArrayInput const&)>>{
-            [](UnitInput const& u) { return u.surfaces.size(); },
-            [](RectArrayInput const&) { return size_type{0}; }};
-
-    auto GetNumVolumes
-        = Overload<std::function<size_type(UnitInput const&)>,
-                   std::function<size_type(RectArrayInput const&)>>{
-            [](UnitInput const& u) { return u.volumes.size(); },
-            [](RectArrayInput const&) { return size_type{0}; }};
+    auto get_num_surfaces
```

Comment:
The template parameter you removed are actually necessary, at least for with the deduction guide as I have it. Without these parameters the build fails:

```
/usr/include/c++/8/variant:835:43: error: invalid conversion from ‘std::__success_type<unsigned int>::type (*)(celeritas::{anonymous}::Overload<celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::UnitInput&)>, celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::RectArrayInput&)> >&, const std::variant<celeritas::UnitInput, celeritas::RectArrayInput>&)’ {aka ‘unsigned int (*)(celeritas::{anonymous}::Overload<celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::UnitInput&)>, celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::RectArrayInput&)> >&, const std::variant<celeritas::UnitInput, celeritas::RectArrayInput>&)’} to ‘long unsigned int (*)(celeritas::{anonymous}::Overload<celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::UnitInput&)>, celeritas::OrangeParams::OrangeParams(celeritas::OrangeInput)::<lambda(const celeritas::RectArrayInput&)> >&, const std::variant<celeritas::UnitInput, celeritas::RectArrayInput>&)’ [-fpermissive]
       { return _Array_type{&__visit_invoke}; }
```
My theory is that the deduction

Reply:
I think that's a bug in GCC8. From the error message it looks like it's because the integer return types are different and it probably isn't converting to a common type. I'll push a more elegant fix than wrapping the lambdas in a `std::function` (which has extra overhead).

Reply:
Yes my understanding is that they all must return the same type